src/core/ext/transport/chttp2/transport/hpack_encoder.cc

```diff
@@ -255,22 +263,20 @@ static void add_key_with_index(grpc_chttp2_hpack_compressor* c,
 
 /* add an element to the decoder table */
 static void add_elem_with_index(grpc_chttp2_hpack_compressor* c,
-                                grpc_mdelem elem, uint32_t new_index) {
+                                grpc_mdelem elem, uint32_t new_index,
+                                uint32_t elem_hash, uint32_t key_hash) {
   if (new_index == 0) {
     return;
   }
-  GPR_ASSERT(GRPC_MDELEM_IS_INTERNED(elem));
-
-  uint32_t key_hash = grpc_slice_hash(GRPC_MDKEY(elem));
-  uint32_t value_hash = grpc_slice_hash(GRPC_MDVALUE(elem));
-  uint32_t elem_hash = GRPC_MDSTR_KV_HASH(key_hash, value_hash);
+  GPR_DEBUG_ASSERT(GRPC_MDELEM_IS_INTERNED(elem));
 
   /* Store this element into {entries,indices}_elem */
-  if (grpc_mdelem_eq(c->entries_elems[HASH_FRAGMENT_2(elem_hash)], elem)) {
+  if (grpc_mdelem_both_interned_eq(c->entries_elems[HASH_FRAGMENT_2(elem_hash)],
```

Comment:
How do you know both mdelems are interned?

Reply:
For the second parameter, elem in add_elem_with_index(c, elem, new_idx, elem_hash, key_hash):

add_elem_with_index is called by: add_elem(c, elem, elem_sz, elem_hash, key_hash)

add_elem() has two callsites, both in: hpack_enc(c, elem, st).

in hpack_enc(): we calculate the variables "elem_interned" and "should_add_elem". Should_add_elem checks if elem is interned as a requirement. Thus, elem is interned.

Now, the 2nd parameter elem is interned is compared with the first parameter: something that is already in the cuckoo hash we're maintaining here. There are only two candidates for valid things that can be in the hash: an empty mdelem  or something that was added via a previous call to add_elem(). Thus, we are guaranteed that the first MD parameter is either interned, or null - both are valid candidates for the short-circuited check.

I think it is fine to add GPR_DEBUG_ASSERT() verifying that this is the case, to guard against people accidentally breaking these assumptions in the future.